### PR TITLE
Update documentation tooling

### PR DESCRIPTION
**Pull request by utzig**
Update documentation build tools to latest versions: Breathe==4.20.0 (released last week) and Sphinx>=3.2.0.

Since the long standing bug that disallowed using the C domain was fixed recently, switch to using it by updating all :cpp: references to equivalent :c: references. This way only the C parser in Sphinx is used and there is one can just use :c: with the type of the identifier being referenced (struct, type, func, enum, member, etc).

### CODE REVIEW

File: .github/workflows/doc-build.yml [hunk 1]
```diff
@@ -30,8 +30,8 @@ jobs:
     - name: install-pip
       run: |
         pip3 install setuptools
-        pip3 install 'breathe>=4.9.1,<4.15.0' 'docutils>=0.14' \
-                     'sphinx>=1.7.5,<3.0' sphinx_rtd_theme sphinx-tabs \
+        pip3 install 'breathe>=4.20.0' 'docutils>=0.16' \
```
Comment (nashif) [inline .github/workflows/doc-build.yml:33]: maybe just use `pip3 install -r scripts/requirements-doc.txt`, this way we do not have to maintain the versions twice and get out of sync.

File: .github/workflows/doc-publish.yml [hunk 2]
```diff
@@ -61,8 +61,8 @@ jobs:
     - name: install-pip
       run: |
         pip3 install setuptools
-        pip3 install 'breathe>=4.9.1,<4.15.0' 'docutils>=0.14' \
-                     'sphinx>=1.7.5,<3.0' sphinx_rtd_theme sphinx-tabs \
+        pip3 install 'breathe>=4.20.0' 'docutils>=0.16' \
```
Comment (nashif) [inline .github/workflows/doc-publish.yml:64]: ditto

**Review by nashif — CHANGES_REQUESTED** [review]

File: .github/workflows/doc-build.yml (hunk 1 above)
Comment (utzig) [reply .github/workflows/doc-build.yml]: done!

File: .github/workflows/doc-publish.yml (hunk 2 above)
Comment (utzig) [reply .github/workflows/doc-publish.yml]: ditto!

**Review by nashif — APPROVED** [review]

File: scripts/requirements-doc.txt [hunk 3]
```diff
@@ -1,8 +1,8 @@
 # DOC: used to generate docs
 
-breathe>=4.9.1,<4.15.0
-docutils>=0.14
-sphinx>=1.7.5,<3.0
+breathe>=4.20.0
+docutils>=0.16
+sphinx>=3.2.0
```
Comment (chrta) [inline scripts/requirements-doc.txt:5]: Only requiring >=3.2.0 for sphinx seems dangerous to me. In the past sphinx changed a lot in their api that broke compatibility with extensions and themes when updating major versions.

Comment (utzig) [reply scripts/requirements-doc.txt:5]: What is your suggestion?

Comment (chrta) [reply scripts/requirements-doc.txt:5]: Maybe `sphinx>=3.2.0,<4.0` would be good for now.

Comment (utzig) [reply scripts/requirements-doc.txt:5]: Ok, the latest release is `3.2.1` and there are no development releases with updated major versions, so it's really hard to know right now if they'll be compatible or not, but I will update as suggested.

Comment (de-nordic) [reply scripts/requirements-doc.txt]: I do not think this is an issue. If `sphinx` follows semantic versioning (https://semver.org/) then change in major version is exactly doing what might have been expected.

Comment (de-nordic) [reply scripts/requirements-doc.txt]: I have misunderstood the problem. Sorry.

**Comment by utzig** [comment]
I removed the DNM now because I am fairly happy with the results when compared to the official docs. I've not seen a single link broken that was not already broken before, and this PR is a huge improvement in this regard, eg, check this section: https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#thread-resource-pools. Every API link is broken there and almost all of them works on this PR. This also applies to every pages touched by this PR that I reviewed manually.

File: scripts/requirements-doc.txt (hunk 3 above)
Comment (utzig) [reply scripts/requirements-doc.txt]: I changed it to what @chrta suggested; my personal opinion would be to not set upper limits right now, but they won't really affect us for a few months at least.

**Review by chrta — CHANGES_REQUESTED** [review]
I am currently building the doc to approve this later. I noticed that the version information in zephyr/doc/README.rst should also be updated, see https://docs.zephyrproject.org/latest/README.html#installing-the-documentation-processors

Edit: The generated doc looks good :+1:

**Comment by utzig** [comment]
> I am currently building the doc to approve this later. I noticed that the version information in zephyr/doc/README.rst should also be updated, see https://docs.zephyrproject.org/latest/README.html#installing-the-documentation-processors

Done; I squashed the update into the first commit that updates the tooling.

**Review by chrta — APPROVED** [review]

**Pull request merged**